### PR TITLE
Migrate projects --> Taxonomy topics

### PR DESCRIPTION
**Pull request by goose-life**
will close https://github.com/laws-africa/indigo-lawsafrica/issues/1036

image

### CODE REVIEW

**Comment by goose-life** [comment]
I've checked the docs for adding a M2M relationship (again), and the error message isn't making any sense to me – I'm hoping it's something minor I've overlooked?

File: indigo_api/migrations/0028_backfill_project_taxonomies.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+        priority_root = high_priority if project.priority else low_priority
+        project_taxonomy = TaxonomyTopic.add_child(priority_root, name=new_title, description=project.description)
+
+        # add project taxonomy to all related works
+        for task in project.tasks.all():
+            if task.work:
+                work = task.work
+                # TODO: neither of these options are working
+                # TypeError: 'TaxonomyTopic' instance expected, got <TaxonomyTopic: South Africa – proooj>
+                work.taxonomy_topics.add(project_taxonomy)
```
Comment (longhotsummer) [inline indigo_api/migrations/0028_backfill_project_taxonomies.py:37]: You're mixing migration-based models and "concrete" models. I don't think Django likes that.

**Comment by actlikewill** [comment]
@goose-life Ah, I think the model methods to manipulate the tree such as `add_root` in the TaxonomyTopic model are not available in a migrations script. Migrations don't have the application context within them. I think you can use go with a management command

File: indigo_api/fixtures/taxonomy_topics.json [hunk 2]
```diff
@@ -1,9 +1,9 @@
 [
   {
     "model": "indigo_api.taxonomytopic",
-    "pk": 1,
+    "pk": 998,
     "fields": {
-      "path": "0001",
+      "path": "0999",
```
Comment (longhotsummer) [inline indigo_api/fixtures/taxonomy_topics.json:6]: If these do change, I think this must be `0998` to match the pk.

Comment (longhotsummer) [reply indigo_api/fixtures/taxonomy_topics.json:6]: but i'm really just guessing

File: indigo_api/migrations/0028_backfill_project_taxonomies.py (hunk 1 above)
Comment (goose-life) [reply indigo_api/migrations/0028_backfill_project_taxonomies.py]: thanks, yeah this was definitely it!

**Review by longhotsummer — APPROVED** [review]

**Comment by actlikewill** [comment]
@goose-life Were you going to also add the taxonomy_topics to the batch upload form on this PR or are you doing it separately?

**Review by actlikewill — APPROVED** [review]

**Pull request merged**